pandas/core/arrays/datetimelike.py

```diff
@@ -364,17 +490,134 @@ def __getitem__(self, key):
+            raise TypeError(msg.format(scalar=self._scalar_type.__name__,
+                                       typ=type(value).__name__))
+        self._data[key] = value
+        self._maybe_clear_freq()
+
+    def _maybe_clear_freq(self):
+        # inplace operations like __setitem__ may invalidate the freq of
+        # DatetimeArray and TimedeltaArray
+        pass
+
+    def isna(self):
+        return self._isnan
+
+    @property  # NB: override with cache_readonly in immutable subclasses
+    def _isnan(self):
+        """
+        return if each value is nan
+        """
+        return (self.asi8 == iNaT)
+
+    @property  # NB: override with cache_readonly in immutable subclasses
+    def hasnans(self):
+        """
+        return if I have any nans; enables various perf speedups
+        """
+        return bool(self._isnan.any())
+
+    def fillna(self, value=None, method=None, limit=None):
+        if isinstance(value, ABCSeries):
+            value = value.array
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        mask = self.isna()
+
+        if is_array_like(value):
+            if len(value) != len(self):
+                raise ValueError("Length of 'value' does not match. Got ({}) "
+                                 " expected {}".format(len(value), len(self)))
+            value = value[mask]
+
+        if mask.any():
+            if method is not None:
+                if method == 'pad':
+                    func = missing.pad_1d
+                else:
+                    func = missing.backfill_1d
+
+                new_values = func(self._data, limit=limit,
+                                  mask=mask)
+                new_values = type(self)(new_values, dtype=self.dtype)
+            else:
+                # fill with value
+                new_values = self.copy()
+                new_values[mask] = value
+        else:
+            new_values = self.copy()
+        return new_values
+
     def astype(self, dtype, copy=True):
+        # Some notes on cases we don't have to handle:
+        #   1. PeriodArray.astype handles period -> period
+        #   2. DatetimeArray.astype handles conversion between tz.
+        #   3. DatetimeArray.astype handles datetime -> period
+        from pandas import Categorical
+        dtype = pandas_dtype(dtype)
+
         if is_object_dtype(dtype):
             return self._box_values(self.asi8)
-        return super(DatetimeLikeArrayMixin, self).astype(dtype, copy)
-
-    # ------------------------------------------------------------------
-    # ExtensionArray Interface
-    # TODO:
-    #   * _from_sequence
-    #   * argsort / _values_for_argsort
-    #   * _reduce
+        elif is_string_dtype(dtype) and not is_categorical_dtype(dtype):
+            return self._format_native_types()
+            # return Index(self.format(), name=self.name, dtype=object)
+        elif is_integer_dtype(dtype):
+            # we deliberately ignore int32 vs. int64 here.
```

Comment:
Agreed. earlier @jreback said this was historical, since DatetimeIndex.astype(int) -> Index, and we don't have an Int32Index.

There's a practical reason to mirror that here, for 32-bit systems `.astype(int)` would almost certainly lead to overflow.

Reply:
But this is not index but array, so I wouldn't use Index's historical reasons as justification.

> for 32-bit systems .astype(int) would almost certainly lead to overflow.

You don't have that only for datetimes, so that is something in general with numpy to take care of?

Reply:
There's something to be said for DatetimeIndex.astype and DatetimeArray.astype behaving identically though.

Personally, I don't have a preference here. It's an easy fix for `DatetimeIndexOpsMixin` to check for integer dtypes before calling `self._data.astype`.

Reply:
no its not historical, rather I said its impossible. We don't support non -64 bit int types for Index as we don't have an exposed user class to do so. Its not that hard, but would actually break tons of things as we rely upon this upcasting in our tests. I agree we should probably try to do this, but likely better when there is more integration between Index and Array.

Reply:
This is about array, which is certainly possible, since we're returning an ndarray here.

But, I agree with you that this shouldn't be done right now because that would introduce a difference in behavior between DatetimeIndex.astype and DatetimeArray.astype.